### ISSUE TITLE
Why did the w value change significantly in a few days?

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs4anki#187]
**Research**

*Enter an [x] character to confirm the points below:*

- [x] I have checked the [FAQ](https://github.com/open-spaced-repetition/fsrs4anki/wiki/FAQ) and could not find an answer to my question
- [x] I have read the [wiki](https://github.com/open-spaced-repetition/fsrs4anki/wiki) and still felt confused
- [x] I have searched for similar existing questions here

**Question**

I calculated the `w` value for my collection on 6th March. The result was:
[1.0726, 1.5077, 5.1018, -1.2078, -1.1282, 0.0, 1.608, -0.0536, 0.9837, 1.6994, -0.5026, 0.7791, 0.593]

Then, I calculated the `w` value for my collection today. The result was:
[1.0802, 1.5621, 5.0744, -1.152, -1.1262, 0.0, 1.6603, -0.0413, 1.0387, 1.7289, -0.4773, 0.7798, 0.6169]

Does this seem to be normal?
With the new `w` value, the number of cards due in the coming 10 days dropped to less than half that with the old `w` value.

Note that during the period from 6th March (actually, from even before that) till day before yesterday, I was working through the backlog which FSRS created for me after I rescheduled all my cards using the helper.

**Comment by L-M-Sherlock** [comment]
Could you reschedule with the old parameters and check the number of due cards?

**Comment by user1823** [comment]
- With the old parameters, the number of due cards for the next 10 days is about 130 cards per day.
- With the new parameters, the number of due cards for the next 10 days is about 60 cards per day.

**Comment by user1823** [comment]
> Do you train the new weights from the initial weights?

What does this mean?

What I did was as follows:
On 6th March, I exported my cards and uploaded them to the optimizer to generate the `w` value. Then, I changed the `w` value in the custom scheduler to that given by the optimizer. And then, I rescheduled all my cards using the helper.

Yesterday, I did the same.

However, after I thought that something was wrong, I changed the `w` value back to the one given by the optimizer on 6th March and then rescheduled again.

**Comment by L-M-Sherlock** [comment]
How many of your review logs were used to generate the w value?

**Comment by user1823** [comment]
- On 6th March, the deck contained 8082 cards with a total of 72,260 reviews.
- Yesterday, the deck contained 8181 cards with a total of 73,602 reviews.

For any other further questions that you may have, I am attaching a copy of the deck as on the two dates concerned.

[6th March.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/10980229/6th.March.zip)
[14th March.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/10980227/14th.March.zip)

**Comment by L-M-Sherlock** [comment]
Thanks for the data! I will do some research this week.

**Comment by cjdduarte** [comment]
> > Do you train the new weights from the initial weights?
> 
> What does this mean?
> 
> What I did was as follows: On 6th March, I exported my cards and uploaded them to the optimizer to generate the `w` value. Then, I changed the `w` value in the custom scheduler to that given by the optimizer. And then, I rescheduled all my cards using the helper.
> 
> Yesterday, I did the same.
> 
> However, after I thought that something was wrong, I changed the `w` value back to the one given by the optimizer on 6th March and then rescheduled again.

@user1823 What version of your Anki?

**Comment by user1823** [comment]
> What version of your Anki?
2.1.60

Though, I did my reviews on AnkiDroid 2.16alpha94.

**Comment by xjqm-tao** [comment]
After reviewing for a period of time using the FSRS algorithm, will it affect the accuracy of parameter optimization?

**Comment by user1823** [comment]
I generated the `w` value again today. The result was:
[1.0725, 1.5855, 5.1263, -1.2133, -1.1616, 0.0044, 1.6274, -0.059, 1.0025, 1.7491, -0.4564, 0.7837, 0.5864]

This yielded a number of due cards that was similar to that was similar to the one yielded by the `w` value I obtained on 6th March. So, I think that the `w` value generated on 14th March was invalid. 

So, the question now becomes: Why did the optimizer give an erroneous result on 14th March?

My deck as of today: [18th March.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/11008624/18th.March.zip)

**Comment by L-M-Sherlock** [comment]
Sorry for this later reply. I want to reproduce the training process with your 6th March data. I need the values of `timezone`, `next_day_starts_at`, and `revlog_start_date`.

**Comment by user1823** [comment]
- `timezone = 'Asia/Calcutta'`
- `next_day_starts_at = 4`
- `revlog_start_date = "2006-10-05"`

**Comment by L-M-Sherlock** [comment]
I guess the optimizer was updated between your two optimizations.

**Comment by user1823** [comment]
> I guess the optimizer was updated between your two optimizations.

It is not the case as is evident in the following screenshot:
![Screenshot_20230409-113435~2](https://user-images.githubusercontent.com/92206575/230757397-1232846e-8641-4a55-8f69-e946f0fff10e.png)

**Comment by L-M-Sherlock** [comment]
So, the extra revlogs recorded between 6th March and 15th March caused a significant change.

**Comment by user1823** [comment]
And then the revlogs between 14th & 18th March reversed the change?

This makes me think that the optimizer assigned a greater significance to the ~1300 reviews between 6th and 14th March as compared to the ~72,300 reviews that were done before 6th March.

**Comment by L-M-Sherlock** [comment]
Maybe increasing the number of epochs and adding adaptive learning rates would help.

**Comment by user1823** [comment]
Given that optimization is something that users don't need to do very frequently, it makes sense to make it more accurate even if it causes the optimizer to take (slightly) more time to generate the parameters.

**Comment by L-M-Sherlock** [comment]
OK. It could be better to allow professional users to set the hyper-parameters easily.

**Comment by user1823** [comment]
> It could be better to allow professional users to set the hyper-parameters easily.
Yes. But, I think that the default values should also be changed slightly so as to achieve a better output while still minimizing the increase in the optimization time.

**Comment by user1823** [comment]
Please fix this issue as soon as possible.

On 9th April, I got the following parameters:
`[1.0409, 1.623, 5.0558, -1.1913, -1.0958, 0.0099, 1.6757, -0.0485, 1.0494, 1.7605, -0.4464, 0.7638, 0.6442]`

On that day, I updated the parameters in my custom scheduling code and rescheduled with the helper.

Today (19th April), I generated the parameters again. I got the following parameters:
`[1.0696, 1.6707, 5.0874, -0.9888, -0.9358, 0.0001, 1.5616, -0.0436, 0.949, 1.8558, -0.3474, 0.6083, 0.7671]`

Rescheduling all cards with these parameters gave me a huge backlog of 1206 cards.

**Comment by L-M-Sherlock** [comment]
I have added the adaptive learning rate `Adam`. And I also added `scheduler = torch.optim.lr_scheduler.OneCycleLR(optimizer, max_lr=lr, total_steps=epoch_len * n_epoch)`.

**Comment by user1823** [comment]
> It could be better to allow professional users to set the hyper-parameters easily.

With the latest update (the code being hidden away in a separate file), it becomes even more important to add an option to easily change the n_epoch and learning rate (and probably other factors also).

**Comment by user1823** [comment]
I think that this behavior might suggest that the model is overfitted.

> In mathematical modeling, overfitting is "the production of an analysis that corresponds too closely or exactly to a particular set of data and may therefore fail to fit to additional data or predict future observations reliably".

> The essence of overfitting is to have unknowingly extracted some of the residual variation (i.e., the [noise](https://en.wikipedia.org/wiki/Statistical_noise)) as if that variation represented underlying model structure.

> over-fitting occurs when a model begins to "memorize" training data rather than "learning" to generalize from a trend.

> As an extreme example, if the number of parameters is the same as or greater than the number of observations, then a model can perfectly predict the training data simply by memorizing the data in its entirety. Such a model, though, will typically fail severely when making predictions.

Source: https://en.wikipedia.org/wiki/Overfitting

Possibly, the optimizer tries to fit the original data too closely that the parameters change greatly when the optimizer is fed newer data.

By the way, I don't know anything about mathematical modeling and this was just a guess.

**Comment by user1823** [comment]
Let me share three sets of parameters generated on different dates, all with n_epoch = 3.

19th April:
`[1.0696, 1.6707, 4.9898, -1.2855, -1.1631, 0.0, 1.7095, -0.0887, 1.0651, 1.71, -0.4937, 0.8296, 0.4661]`

24th April:
`[1.0574, 1.6845, 5.0361, -1.3685, -1.2342, 0.0002, 1.7343, -0.0874, 1.0907, 1.7128, -0.4931, 0.8331, 0.4939]`

5th May:
`[1.0452, 1.6989, 4.9877, -1.26, -1.1786, 0.0, 1.7037, -0.0941, 1.0638, 1.7052, -0.4985, 0.8364, 0.5216]`

Here, you can see how the parameters are changing continuously.

**Comment by L-M-Sherlock** [comment]
Does this issue still exist?

**Comment by L-M-Sherlock** [comment]
Due to the inactivity, I close this issue.

**Comment by user1823** [comment]
Well, this issue still exists.

Let me share a list of parameters generated on different dates and the number of due cards I obtain after rescheduling my today's collection with those parameters.

31st May
[1.3589, 1.9487, 4.4756, -2.0243, -1.8267, 0.0004, 1.8586, -0.1469, 1.2332, 1.7557, -0.5268, 0.8822, 0.2489]
Number of due cards = 0

15th June
[1.5382, 2.558, 4.066, -2.0442, -1.9435, 0.05, 1.7368, -0.2021, 1.1172, 1.8003, -0.5583, 0.8502, 0.2502]
Number of due cards = 898

24th June
[1.6117, 2.6584, 4.0081, -2.0834, -1.9722, 0.05, 1.7121, -0.2091, 1.14, 1.7954, -0.5468, 0.8486, 0.1485]
Number of due cards = 972

Note: The number of due cards with parameters of 31st May is zero because I am currently using those parameters and I have completed today's reviews.

**Comment by user1823** [comment]
By the way, would you suggest me to change the parameters in the scheduler to those generated today?

**Comment by user1823** [comment]
Another observation: The increase in minimum value of w[5] was supposed to decrease the workload (by preventing ease hell). But, here, it is increasing the workload.
